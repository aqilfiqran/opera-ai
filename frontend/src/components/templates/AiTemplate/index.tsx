'use client';

import Container from '@/components/atoms/Container';
import Text from '@/components/atoms/Text';
import InputQuestion from '@/components/molecules/InputQuestion';
import TextMarkdown from '@/components/molecules/TextMarkdown';
import { useAskAi } from '@/hooks/useAskAi';
import { cn } from '@/lib/utils';
import { useState } from 'react';

export default function AiTemplate() {
  const [question, setQuestion] = useState('');
  const [answer, setAnswer] = useState('');
  const { trigger, loading } = useAskAi();

  const handleAskQuestion = async () => {
    trigger(question, {
      onSuccess: response => {
        setAnswer(response.data);
      },
      onError: error => {
        console.error('Error fetching AI response:', error);
        setAnswer('');
      },
    });
  };

  return (
    <Container className="flex flex-col gap-3 py-5">
      <div className="flex flex-col">
        <Text variant="xl" as="h2">
          Opera AI
        </Text>
        <Text variant="md" as="h3" className="font-normal text-neutral-500">
          AI-powered sales assistant for Opera. Ask me anything about Opera, sales, or AI!
        </Text>
      </div>

      {!!answer && (
        <section
          className={cn(
            'flex flex-col gap-1',
            // transition from top to bottom when answer is loaded
            'animate-fade-in-down duration-500 ease-in-out'
          )}
        >
          <Text variant="sm" className="text-primary font-bold">
            Response
          </Text>
          <div className="relative rounded-lg border border-neutral-300 bg-white p-4 shadow-md">
            {/* markdown format support */}
            <TextMarkdown text={answer} />
          </div>
        </section>
      )}

      <div className="fixed right-4 bottom-5 left-4 flex justify-center">
        <InputQuestion
          value={question}
          onChange={e => setQuestion(e.target.value)}
          onValueSubmit={handleAskQuestion}
          loading={loading}
        />
      </div>
    </Container>
  );
}
