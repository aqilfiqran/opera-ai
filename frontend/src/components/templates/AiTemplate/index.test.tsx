import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import AiTemplate from './index';

// Mock the hooks and components
jest.mock('@/hooks/useAskAi', () => ({
  useAskAi: () => ({
    trigger: jest.fn(),
    loading: false,
  }),
}));

jest.mock('@/components/molecules/TextMarkdown', () => ({
  __esModule: true,
  default: ({ text }: { text: string }) => <div data-testid="markdown">{text}</div>,
}));

describe('AiTemplate', () => {
  it('renders correctly in initial state', () => {
    const { asFragment } = render(<AiTemplate />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with answer', () => {
    // Mock useState to return an answer
    const originalUseState = React.useState;
    jest.spyOn(React, 'useState').mockImplementationOnce(() => ['Test question', jest.fn()]);
    jest.spyOn(React, 'useState').mockImplementationOnce(() => ['This is a test answer', jest.fn()]);

    const { asFragment } = render(<AiTemplate />);
    expect(asFragment()).toMatchSnapshot();

    // Restore useState
    jest.spyOn(React, 'useState').mockImplementation(originalUseState);
  });

  it('renders correctly in loading state', () => {
    // Mock useAskAi to return loading=true
    jest.mock('@/hooks/useAskAi', () => ({
      useAskAi: () => ({
        trigger: jest.fn(),
        loading: true,
      }),
    }));

    const { asFragment } = render(<AiTemplate />);
    expect(asFragment()).toMatchSnapshot();
  });
});
