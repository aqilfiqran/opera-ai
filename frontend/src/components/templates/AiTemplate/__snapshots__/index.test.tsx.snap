// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AiTemplate renders correctly in initial state 1`] = `
<DocumentFragment>
  <div
    class="mx-auto w-full px-3 sm:max-w-[540px] md:max-w-[720px] md:px-0 flex flex-col gap-3 py-5"
  >
    <div
      class="flex flex-col"
    >
      <h2
        class="text-mb-xl tracking-2 lg:text-dk-xl font-sans"
      >
        Opera AI
      </h2>
      <h3
        class="tracking-2 lg:text-dk-md font-sans font-normal text-neutral-500"
      >
        AI-powered sales assistant for Opera. Ask me anything about Opera, sales, or AI!
      </h3>
    </div>
    <div
      class="fixed right-4 bottom-5 left-4 flex justify-center"
    >
      <div
        class="flex w-full cursor-text border text-neutral-500 has-[input:disabled]:cursor-not-allowed has-[input:disabled]:bg-neutral-200 has-[input:disabled]:text-neutral-700 has-[input:focus-visible]:ring-primary has-[input:focus-visible]:ring-1 rounded-full border-neutral-500 bg-neutral-200 transition-all duration-200 ease-in-out max-w-xs has-[input:focus-visible]:max-w-full md:has-[input:focus-visible]:max-w-lg"
        role="button"
        tabindex="0"
      >
        <input
          class="text-dark flex font-medium transition-colors disabled:cursor-not-allowed disabled:bg-neutral-200 disabled:text-neutral-400 disabled:placeholder:text-neutral-600 focus-visible:outline-none w-full px-2 py-1.5 rounded-full bg-neutral-200 pl-5 placeholder:text-neutral-400"
          data-slot="input"
          placeholder="ask AI. press '/' to focus"
          type="text"
          value=""
        />
        <div
          class="py-1.5 pr-2"
        >
          <div
            class="flex h-9 w-9 cursor-pointer items-center justify-center rounded-full bg-neutral-500 transition-colors hover:bg-neutral-600"
            role="button"
            tabindex="0"
          >
            <svg
              class="lucide lucide-arrow-up text-neutral-200"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m5 12 7-7 7 7"
              />
              <path
                d="M12 19V5"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AiTemplate renders correctly in loading state 1`] = `
<DocumentFragment>
  <div
    class="mx-auto w-full px-3 sm:max-w-[540px] md:max-w-[720px] md:px-0 flex flex-col gap-3 py-5"
  >
    <div
      class="flex flex-col"
    >
      <h2
        class="text-mb-xl tracking-2 lg:text-dk-xl font-sans"
      >
        Opera AI
      </h2>
      <h3
        class="tracking-2 lg:text-dk-md font-sans font-normal text-neutral-500"
      >
        AI-powered sales assistant for Opera. Ask me anything about Opera, sales, or AI!
      </h3>
    </div>
    <div
      class="fixed right-4 bottom-5 left-4 flex justify-center"
    >
      <div
        class="flex w-full cursor-text border text-neutral-500 has-[input:disabled]:cursor-not-allowed has-[input:disabled]:bg-neutral-200 has-[input:disabled]:text-neutral-700 has-[input:focus-visible]:ring-primary has-[input:focus-visible]:ring-1 rounded-full border-neutral-500 bg-neutral-200 transition-all duration-200 ease-in-out max-w-xs has-[input:focus-visible]:max-w-full md:has-[input:focus-visible]:max-w-lg"
        role="button"
        tabindex="0"
      >
        <input
          class="text-dark flex font-medium transition-colors disabled:cursor-not-allowed disabled:bg-neutral-200 disabled:text-neutral-400 disabled:placeholder:text-neutral-600 focus-visible:outline-none w-full px-2 py-1.5 rounded-full bg-neutral-200 pl-5 placeholder:text-neutral-400"
          data-slot="input"
          placeholder="ask AI. press '/' to focus"
          type="text"
          value=""
        />
        <div
          class="py-1.5 pr-2"
        >
          <div
            class="flex h-9 w-9 cursor-pointer items-center justify-center rounded-full bg-neutral-500 transition-colors hover:bg-neutral-600"
            role="button"
            tabindex="0"
          >
            <svg
              class="lucide lucide-arrow-up text-neutral-200"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m5 12 7-7 7 7"
              />
              <path
                d="M12 19V5"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AiTemplate renders correctly with answer 1`] = `
<DocumentFragment>
  <div
    class="mx-auto w-full px-3 sm:max-w-[540px] md:max-w-[720px] md:px-0 flex flex-col gap-3 py-5"
  >
    <div
      class="flex flex-col"
    >
      <h2
        class="text-mb-xl tracking-2 lg:text-dk-xl font-sans"
      >
        Opera AI
      </h2>
      <h3
        class="tracking-2 lg:text-dk-md font-sans font-normal text-neutral-500"
      >
        AI-powered sales assistant for Opera. Ask me anything about Opera, sales, or AI!
      </h3>
    </div>
    <section
      class="flex flex-col gap-1 animate-fade-in-down duration-500 ease-in-out"
    >
      <p
        class="tracking-2 lg:text-dk-sm font-sans text-primary font-bold"
      >
        Response
      </p>
      <div
        class="relative rounded-lg border border-neutral-300 bg-white p-4 shadow-md"
      >
        <div
          data-testid="markdown"
        >
          This is a test answer
        </div>
      </div>
    </section>
    <div
      class="fixed right-4 bottom-5 left-4 flex justify-center"
    >
      <div
        class="flex w-full cursor-text border text-neutral-500 has-[input:disabled]:cursor-not-allowed has-[input:disabled]:bg-neutral-200 has-[input:disabled]:text-neutral-700 has-[input:focus-visible]:ring-primary has-[input:focus-visible]:ring-1 rounded-full border-neutral-500 bg-neutral-200 transition-all duration-200 ease-in-out max-w-xs has-[input:focus-visible]:max-w-full md:has-[input:focus-visible]:max-w-lg"
        role="button"
        tabindex="0"
      >
        <input
          class="text-dark flex font-medium transition-colors disabled:cursor-not-allowed disabled:bg-neutral-200 disabled:text-neutral-400 disabled:placeholder:text-neutral-600 focus-visible:outline-none w-full px-2 py-1.5 rounded-full bg-neutral-200 pl-5 placeholder:text-neutral-400"
          data-slot="input"
          placeholder="ask AI. press '/' to focus"
          type="text"
          value="Test question"
        />
        <div
          class="py-1.5 pr-2"
        >
          <div
            class="flex h-9 w-9 cursor-pointer items-center justify-center rounded-full bg-neutral-500 transition-colors hover:bg-neutral-600"
            role="button"
            tabindex="0"
          >
            <svg
              class="lucide lucide-arrow-up text-neutral-200"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m5 12 7-7 7 7"
              />
              <path
                d="M12 19V5"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
